File: ios/building/RMFBuilding.m
//
//  RMFBuilding.m
//  react-native-map4d-map
//
//  Created by Huy Dang on 6/10/25.
//  Copyright © 2025 IOTLink. All rights reserved.
//

#import "RMFBuilding.h"
#import "RMFEventResponse.h"

@implementation RMFBuilding

- (instancetype)init {
  self = [super init];
  if (self) {
    _map4dBuilding = [[RMFBuildingMap4d alloc] init];
    _map4dBuilding.reactBuilding = self;

    _name = _map4dBuilding.name;
    _coordinate = _map4dBuilding.position;
    _modelUrl = _map4dBuilding.model;
    _textureUrl = _map4dBuilding.texture;
    _scale = _map4dBuilding.scale;
    _bearing = _map4dBuilding.bearing;
    _elevation = _map4dBuilding.elevation;
    _selected = _map4dBuilding.selected;
    _touchable = _map4dBuilding.userInteractionEnabled;
    _visible = !_map4dBuilding.isHidden;
    _userData = nil;
  }
  return self;
}

- (void)setMapView:(RMFMapView *)mapView {
  _map4dBuilding.map = mapView;
}

- (void)didTapAtPixel:(CGPoint)pixel {
  if (self.onPress) {
    CLLocationCoordinate2D tapLocation = [_map4dBuilding.map.projection coordinateForPoint:pixel];
    NSMutableDictionary* response = [NSMutableDictionary dictionaryWithDictionary:[RMFEventResponse fromCoordinate:tapLocation
                                                                                                             pixel:pixel]];
    response[@"building"] = @{
      kRMFLatLngCoordinateResponseKey: [RMFEventResponse fromCoordinate:_map4dBuilding.position],
      @"userData": _userData != nil ? _userData : @{}
    };
    response[@"action"] = @"building-press";
    
    self.onPress(response);
  };
}

- (void)setName:(NSString *)name {
  _name = name;
  _map4dBuilding.name = name;
}

- (void)setCoordinate:(CLLocationCoordinate2D)coordinate {
  _coordinate = coordinate;
  _map4dBuilding.position = coordinate;
}

- (void)setModelUrl:(NSString *)modelUrl {
  _modelUrl = modelUrl;
  _map4dBuilding.model = modelUrl;
}

- (void)setTextureUrl:(NSString *)textureUrl {
  _textureUrl = textureUrl;
  _map4dBuilding.texture = textureUrl;
}

- (void)setScale:(double)scale {
  _scale = scale;
  _map4dBuilding.scale = scale;
}

- (void)setBearing:(CGFloat)bearing {
  _bearing = bearing;
  _map4dBuilding.bearing = bearing;
}

- (void)setElevation:(double)elevation {
  _elevation = elevation;
  _map4dBuilding.elevation = elevation;
}

- (void)setSelected:(BOOL)selected {
  _selected = selected;
  _map4dBuilding.selected = selected;
}

- (void)setTouchable:(BOOL)touchable {
  _touchable = touchable;
  _map4dBuilding.userInteractionEnabled = touchable;
}

- (void)setVisible:(BOOL)visible {
  _visible = visible;
  _map4dBuilding.isHidden = !visible;
}

@end

File: android/src/main/java/com/reactnativemap4dmap/RMFBuilding.java
package com.reactnativemap4dmap;

import android.content.Context;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

import java.util.ArrayList;
import java.util.List;

import vn.map4d.map.annotations.MFBuilding;
import vn.map4d.map.annotations.MFBuildingOptions;
import vn.map4d.map.core.Map4D;
import vn.map4d.types.MFLocationCoordinate;

public class RMFBuilding extends RMFFeature {

  private MFBuilding building;

  private String name;

  private MFLocationCoordinate coordinate;

  private String model;

  private String texture;

  private double scale;

  private double bearing;

  private float elevation;

  private boolean selected;

  private boolean visible;

  private boolean touchable;

  private String userData;

  public RMFBuilding(Context context) {
    super(context);
    name = "";
    model = "";
    texture = "";
    coordinate = new MFLocationCoordinate(0, 0);
    scale = 1.0;
    bearing = 0.0;
    elevation = 0.f;
    selected = false;
    visible = true;
    touchable = true;
    userData = null;
  }

  public void setName(String name) {
    this.name = name;
    if (building != null) {
      building.setName(this.name);
    }
  }

  public void setCoordinate(ReadableMap data) {
    this.coordinate = new MFLocationCoordinate(data.getDouble("latitude"), data.getDouble("longitude"));
    if (building != null) {
      building.setLocation(this.coordinate);
    }
  }

  public void setModelUrl(String url) {
    this.model = url;
    if (building != null) {
      building.setModel(this.model);
    }
  }

  public void setTextureUrl(String url) {
    this.texture = url;
    if (building != null) {
      building.setTexture(this.texture);
    }
  }

  public void setScale(double scale) {
    this.scale = scale;
    if (building != null) {
      building.setScale(this.scale);
    }
  }

  public void setBearing(double bearing) {
    this.bearing = bearing;
    if (building != null) {
      building.setBearing(this.bearing);
    }
  }

  public void setElevation(float elevation) {
    this.elevation = elevation;
    if (building != null) {
      building.setElevation(this.elevation);
    }
  }

  public void setSelected(boolean selected) {
    this.selected = selected;
    if (building != null) {
      building.setSelected(this.selected);
    }
  }

  public void setVisible(boolean visible) {
    this.visible = visible;
    if (building != null) {
      building.setVisible(this.visible);
    }
  }

  public void setTouchable(boolean touchable) {
    this.touchable = touchable;
    if (building != null) {
      building.setTouchable(this.touchable);
    }
  }

  public void setUserData(ReadableMap userData) {
    this.userData = userData.toString();
    if (building != null) {
      building.setUserData(this.userData);
    }
  }

  @Override
  public void addToMap(Map4D map) {
    this.building = map.addBuilding(getOptions());
  }

  @Override
  public void removeFromMap(Map4D map) {
    if (building == null) {
      return;
    }
    building.remove();
    building = null;
  }

  @Override
  public Object getFeature() {
    return building;
  }

  public MFBuildingOptions getOptions() {
    MFBuildingOptions options = new MFBuildingOptions();
    options.name(name);
    options.location(coordinate);
    options.model(model);
    options.texture(texture);
    options.scale(scale);
    options.bearing(bearing);
    options.elevation(elevation);
    options.visible(visible);
    options.touchable(touchable);
    options.userData(userData);
    return options;
  }
}
